pytorch_lightning/trainer/connectors/logger_connector/result_new.py

```diff
@@ -0,0 +1,499 @@
+# Copyright The PyTorch Lightning team.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+from collections.abc import Generator
+from dataclasses import dataclass, field
+from functools import partial, wraps
+from typing import Any, Callable, Dict, Iterable, Mapping, Optional, Tuple, Union
+
+import torch
+from torchmetrics import Metric
+
+from pytorch_lightning.trainer.connectors.logger_connector.fx_validator import FxValidator
+from pytorch_lightning.utilities import rank_zero_warn
+from pytorch_lightning.utilities.apply_func import apply_to_collection, apply_to_collections
+from pytorch_lightning.utilities.device_dtype_mixin import DeviceDtypeModuleMixin
+from pytorch_lightning.utilities.enums import LightningEnum
+from pytorch_lightning.utilities.exceptions import MisconfigurationException
+from pytorch_lightning.utilities.metrics import metrics_to_scalars
+
+# re-define the ones from pytorch_lightning.utilities.types without the `Number` type
+_METRIC = Union[Metric, torch.Tensor]
+_METRIC_COLLECTION = Union[_METRIC, Mapping[str, _METRIC]]
+
+
+class MetricSource(LightningEnum):
+    CALLBACK = "callback"
+    PBAR = "pbar"
+    LOG = "log"
```

Comment:
are these metric sources, or are they destinations? 

in the future, i can imagine we want to configure multiple loggers, but want to log metrics to only some of them. how could this be extended for that use case?

Reply:
> are these metric sources, or are they destinations?

Mmm depends on how you think about it.

Do you `self.log(prog_bar=True)` into the progress bar metrics (destination) or are the progress bar metrics gathered from the `prog_bar=True` logged values (source)?

If you have an idea for a less confusing name, we can rename without issues 👍 

> in the future, i can imagine we want to configure multiple loggers, but want to log metrics to only some of them. how could this be extended for that use case?

In that case, I think it's easier for us to filter at the logger level. `_Metadata` could have a field for the target logger